### ISSUE TITLE
React reusable component example

### ISSUE THREAD

**Issue by gojutin** [tw-in-js/twind#44]
I created an example reusable React component using Twind and thought I’d share the link for anyone interested. If you have any thoughts on how to improve it, please share!

CodeSandbox Link: https://codesandbox.io/s/twind-react-typescript-wmm6t

- Written in React/TypeScript
- Styled with Twind
- Exported type definitions with JSDoc comments for easy Storybook integration
- Uses forward ref to allow for setting ref on instance
- Allows for className overriding via the `override` variant
- Example usage in App component
- Logged classNames for convenience

**Comment by sastan** [comment]
That looks just amazing!

I see you used a lot of twind features. Really great.

Things i really like:

- `bg(${variantColor}(600 700(hover:& focus:&)))` – that's something 👍 
- using override for customization
- using props for dynamic directive names
- using grouping 
- logging the class names

Just a few things if you are open for it:

- the override variant is included in twind since v0.10.0 – no need to define in setup

The next two are just to show what may be used for optimal performance.

- the `sizeClassNames` can be optimized to be inline function which can be cached

  ```js
    const sizeClassNames = {
      sm: ({ tw }) => tw`text-xs py-1`,
      md: ({ tw }) => tw`text-sm py-1`,
      lg: ({ tw }) => tw`text-lg py-2`,
      xl: ({ tw }) => tw`text-2xl py-2`
    };
  ```

- you could inline `bgClassNames` into `classNames` which would prevent parsing it every time

  ```js
      // Get the text size based on the "size" prop
      const buttonSizeClassNames = sizeClassNames[size];
      // Create the full className, passing the instance-level classNames to the "override" variant
      const classNames = tw`
      w(full md:auto) 
      bg(${variantColor}(600 700(hover:& focus:&)))
      ${buttonSizeClassNames}
      text(sm white uppercase) 
      px-4 
      rounded-${round ? "full" : "lg"} 
      border-none 
      transition-colors 
      duration-300
      override:(${className})
    `;
  ```

This is really good. Thanks for sharing.

Would you be interested in testing/trying the twind/styled PR?

**Comment by sastan** [comment]
If you are OK with it i'd like to transfer this to discussions "Show & Tell" where others can easily find it?

**Comment by sastan** [comment]
After seeing this application we may need to support some kind of comment syntax with template literals...

**Comment by gojutin** [comment]
Thanks for all the feedback and happy to share! I really enjoy using Twind.

As far as inlining the `bgClassNames`…I originally tried that and ran into these errors: `Error: [UNKNOWN_THEME_VALUE] {"key":"backgroundColor.600”}`. I just tried again and am seeing the same thing.

I’ve made the other changes that you recommended. I appreciate the tip about caching with the inline functions 👍.  

And, yes, I’d be happy to try out the styled API. I skimmed through the PR the other day and I’ve actually been looking forward to trying it! And sure, you are welcome to add this example to the “Show & Tell”! And yes, comments would also be great!
